### PR TITLE
[apex] New Rule: AvoidNonRestrictiveQueries

### LINKED ISSUE #635: [apex] New Rule: Avoid soql/sosl queries without a where clause or limit statement
**Rule Set:** Apex, perfomance

**Description:**
When working with very large amounts of data, unfiltered SOQL queries can quickly cause governor limit exceptions 

**Code Sample demonstrating the issue:**

```
public class Something {
	public static void main( String as[] ) {  
		Account[] accs = [ select id from account ];  //Bad
	}
}
```

### PR DESCRIPTION
**Pull request by adangel**
## Describe the PR

- Adds a new rule AvoidNonRestrictiveQueries as suggested back then in #663 
- Ported the rule to PMD 7
- Added more test cases

## Related issues

- Fixes #635
- Refs old PR #663 

## Ready?

- [ ] Added unit tests for fixed bug/feature
- [ ] Passing all unit tests
- [ ] Complete build `./mvnw clean verify` passes (checked automatically by github actions)
- [ ] Added (in-code) documentation (if needed)

### CODE REVIEW

File: pmd-apex/src/test/resources/net/sourceforge/pmd/lang/apex/rule/performance/xml/AvoidNonRestrictiveQueries.xml [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+
+    <test-code disabled="true">
+        <description>main query with limit, subquery without is OK</description>
+        <expected-problems>0</expected-problems>
+        <code><![CDATA[
+class Query {
+    void method() {
+        Account[] accounts = [ select id, (SELECT Id FROM Contact) from account limit 1 ];
+    }
+}
```
Comment (adangel) [inline pmd-apex/src/test/resources/net/sourceforge/pmd/lang/apex/rule/performance/xml/AvoidNonRestrictiveQueries.xml:107]: @dschach
I've updated the original rule from #663 to see what's possible. I now have some contradictions, maybe you can help?
Should this query be flagged by the rule, because it will exceed governor limits?
Or maybe should we flag this, because the subquery SELECT potentially selects all rows and is therefore a performance issue? (the rule is currently in category performance, so this could make sense...).

This test case basically contradicts the test case in lines 42-45...

**Comment by adangel** [comment]
Note - a related rule is maybe [OperationWithLimitsInLoop](https://docs.pmd-code.org/latest/pmd_rules_apex_performance.html#operationwithlimitsinloop) ...

File: pmd-apex/src/test/resources/net/sourceforge/pmd/lang/apex/rule/performance/xml/AvoidNonRestrictiveQueries.xml (hunk 1 above)
Comment (dschach) [reply pmd-apex/src/test/resources/net/sourceforge/pmd/lang/apex/rule/performance/xml/AvoidNonRestrictiveQueries.xml:107]: @adangel If I were to put the rule in plain English:
1. In the SOQL query, throw away any subqueries - we don't care about those.
2. On the main query, require either the word WHERE or LIMIT.
Therefore, the query here is perfectly fine.

**Comment by dschach** [comment]
@adangel Here is a feature request: Include a property for the rule to ignore it in test classes that do not have `@isTest(seealldata=true)`. 
In a test situation without that annotation (which means 99% of test classses) we create records in the test, so filtered queries are not only unnecessary but sometimes are bad.

**Comment by pmd-test** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="true">Compared to master:<br />
This changeset changes 0 violations,<br />
introduces 7 new violations, 0 new errors and 0 new configuration errors,<br />
removes 0 violations, 0 errors and 0 configuration errors.<br />
<a href="https://github.com/pmd/pmd/actions/runs/9699032641?pr=5089">Download full report as build artifact</a></td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: pmd-apex/src/test/resources/net/sourceforge/pmd/lang/apex/rule/performance/xml/AvoidNonRestrictiveQueries.xml (hunk 1 above)
Comment (adangel) [reply pmd-apex/src/test/resources/net/sourceforge/pmd/lang/apex/rule/performance/xml/AvoidNonRestrictiveQueries.xml:107]: Thanks for the clarification - that makes it easier. The rule should only check the main query.

**Comment by adangel** [comment]
> @adangel Here is a feature request: Include a property for the rule to ignore it in test classes that do not have `@isTest(seealldata=true)`. In a test situation without that annotation (which means 99% of test classses) we create records in the test, so filtered queries are not only unnecessary but sometimes are bad.

Good point, thanks for mentioning this. This also came up back then (see https://github.com/pmd/pmd/pull/663#issuecomment-337582830).

Use [@IsTest(SeeAllData=true)](https://developer.salesforce.com/docs/atlas.en-us.apexcode.meta/apexcode/apex_classes_annotation_isTest.htm#SeeAllDataSection) as an indication seems fine. Do we really need a property? I would say, the rule should by default ignore tests, unless SeeAllData is used...

Hm... ok, I see the problem: Before Summer '24, the tests _always had access to all data_ (see https://developer.salesforce.com/docs/atlas.en-us.apexcode.meta/apexcode/apex_testing_data_access.htm) - so the rule should flag also tests, but only for older versions... (< Summer '24, API 61.0). While we have support for different language versions, AFAIK this feature is not used very much - most likely the default language version will be used which is the latest version... So maybe a property could be useful after all?

**Comment by adangel** [comment]
> Hm... ok, I see the problem: Before Summer '24, the tests _always had access to all data_ (see https://developer.salesforce.com/docs/atlas.en-us.apexcode.meta/apexcode/apex_testing_data_access.htm) - so the rule should flag also tests, but only for older versions... (< Summer '24, API 61.0).

I confused "Summer '24" with API version 24.0 - the documentation says:

> Starting with Apex code saved using Salesforce API version 24.0, test methods don’t have access to pre-existing data in the organization. However, test code saved against Salesforce API version 23.0 and earlier continues to have access to all data in the organization. 

So, that behavioral/default change happened already more than 10 years ago...
And this annotation parameter exists already for more than 10 years.

> So maybe a property could be useful after all?

I think, we don't need a property. We should assume, that one uses Apex API > 24.0 anyways...

File: pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/rule/performance/AvoidNonRestrictiveQueriesRule.java [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+import net.sourceforge.pmd.lang.apex.rule.AbstractApexRule;
+import net.sourceforge.pmd.lang.ast.NodeStream;
+import net.sourceforge.pmd.lang.rule.RuleTargetSelector;
+import net.sourceforge.pmd.reporting.RuleContext;
+
+public class AvoidNonRestrictiveQueriesRule extends AbstractApexRule {
+    private static final Pattern RESTRICTIVE_PATTERN = Pattern.compile("(where )|(limit )", Pattern.CASE_INSENSITIVE);
+    private static final Pattern SELECT_PATTERN = Pattern.compile("(select )", Pattern.CASE_INSENSITIVE);
+    private static final Pattern SUB_QUERY_PATTERN = Pattern.compile("(?i)\\(\\s*select\\s+[^)]+\\)");
+    private static final String SEE_ALL_DATA_ANNOTATION_PARAMETER = "SeeAllData";
```
Comment (adangel) [inline pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/rule/performance/AvoidNonRestrictiveQueriesRule.java:28]: See https://github.com/pmd/pmd/blob/e6397d5d5ce613d57280c04bf97cddd679259e09/pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/ast/ASTAnnotationParameter.java#L13

**Comment by dschach** [comment]
This is great so far. Can we extend it to SOSL, so we'd check for SELECT or FIND?

**Comment by adangel** [comment]
> This is great so far. Can we extend it to SOSL, so we'd check for SELECT or FIND?

@dschach I've updated to rule to consider SOSL as well. Can you have another look at it (especially the test cases)?

**Comment by adangel** [comment]
> Compared to master:
> This changeset changes 0 violations,
> introduces 2 new violations, 0 new errors and 0 new configuration errors,
> removes 0 violations, 0 errors and 0 configuration errors.

These two new violations seem to be false-positives, since they contain a WHERE clause:

* https://github.com/SalesforceLabs/Schedul-o-matic-9000/blob/6b1229ba43b38931fbbab5924bc9b9611d19a786/force-app/main/default/classes/Scheduler.cls#L38
* https://github.com/SalesforceLabs/Schedul-o-matic-9000/blob/6b1229ba43b38931fbbab5924bc9b9611d19a786/force-app/main/default/classes/TestUtils.cls#L108

**Review by dschach — APPROVED** [review]

**Comment by dschach** [comment]
@adangel I really like the tests in this PR. It's great to be able to "filter" this rule based on seealldata=true/false when running queries in test classes. Yes, if false, unrestricted queries are fine. If true... not good.
Should we update some of the language to specify SOSL as well, or just leave that as implied, as we do with other query-related rules? If we leave it, then I'm cool with that and this looks beautiful.

**Comment by adangel** [comment]
@dschach Thanks for the review.

> Should we update some of the language to specify SOSL as well, or just leave that as implied, as we do with other query-related rules?

I've already updated the wording to use "SOQL/SOSL" in the rule description and in the violation message explicitly either SOQL or SOSL. So this should be fine.

I hopefully fixed the two false positives now. Once that is confirmed, I'll merge this for 7.4.0.
